daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala

```diff
@@ -73,11 +73,15 @@ private[preprocessing] final class CommandPreprocessor(compiledPackages: Mutable
       contractId: Value.ContractId,
       choiceId: Ref.ChoiceName,
       argument: Value[Value.AbsoluteContractId],
-  ): speedy.Command.Exercise = {
+  ): (speedy.Command.Exercise, Set[Value.AbsoluteContractId]) = {
     val template = unsafeGetTemplate(templateId)
     val choiceArgType = unsafeGetChoiceArgType(templateId, template, choiceId)
-    val arg = valueTranslator.unsafeTranslateValue(choiceArgType, argument)
-    speedy.Command.Exercise(templateId, SValue.SContractId(contractId), choiceId, arg)
+    val (arg, argCids) = valueTranslator.unsafeTranslateValue(choiceArgType, argument)
```

Comment:
Why do we care about contractIds referred to in a value? Since this can come from the ledger api, anything can be put in there. See @andreaslochbihler-da's [comment](https://github.com/digital-asset/daml/pull/5361#discussion_r417146998).

Reply:
The purpose of this check is the discriminator of the contract created by the transaction are globally fresh, that is they do not conflict with any other contract IDs refered by a transaction. 

For this purpose we introduce the notion of local and global contract ID that slightly differ from input/output contracts. 

A _local contract ID_ is the ID of a contract created in a transaction.  
Note that local contract IDs correspond to the IDs of output contract ID together with those which are created and consumed in the same transaction. 

A _global contract ID_ is contract ID that: 
1 appears in the command itself (notably in the argument of the create, exercise)  
2 the IDs of input contract fetched or looked up by key
3 the IDs referenced in the body of the input contracts. 

Note that global contract do not reference only to IDs of some contract that exists (or has existed) on the ledger  but also any random CID somebody can submit to the ledger API

The freshness check consists to verify the discriminators of the local ID are distinct from the global ID. 

The value here is part of the command (point 1) and should be  consider as a global ID.

Reply:
Great explanation. Would you mind putting this as a comment on `Machine` somewhere?

Reply:
done.